plugins/nf-co2footprint/src/main/nextflow/co2footprint/CarbonIntensity.groovy

```
@@ -0,0 +1,179 @@
+                ${HelperFunctions.bold('────────────────────────────────────────────────────')}
+                """)
+                return realTimeCI
+            } else {
+                log.warn(Markers.unique, "Could not retrieve real-time carbon intensity for ${HelperFunctions.bold(this.location)}.")
+            }
+        } catch (Exception e) {
+            log.error(Markers.unique, "Error retrieving real-time carbon intensity for ${HelperFunctions.bold(this.location)}: ${e.message}")
+        }
+        return null
+    }
+
+    /**
+    * Determines the carbon intensity for a given zone.
+    *
+    * @return A closure that retrieves the carbon intensity value when invoked.
+    */
+    private Closure<Double> getCI() {
+        return { ->
+            def ci
+
+            if (this.location) {
+                this.location = this.location.toUpperCase() // Ensure location is always uppercase
+
+                // Check if the API key is set and is a valid string
+                if (this.apiKey && this.apiKey instanceof String) {
+                    log.info(Markers.unique, "API key is set. Attempting to retrieve real-time carbon intensity.")
+
+                    // Attempt to retrieve real-time carbon intensity
+                    ci = getRealtimeCI()
+                    if (ci != null) {
+                        log.info(Markers.unique, "Using real-time carbon intensity for ${HelperFunctions.bold(this.location)}: ${HelperFunctions.bold(ci.toString())} gCO₂eq/kWh.")
+                        return ci
+                    } else {
+                        log.warn("Could not retrieve real-time carbon intensity for ${HelperFunctions.bold(this.location)}.")
+                    }
+                } else {
+                    log.warn(Markers.unique, "API key is not set or invalid. Skipping real-time carbon intensity retrieval.")
+                }
+
+                // Fallback to the location in the CSV
```

I implemented a `boolean fallbackToDefault` to check whether we want that or would scrap the run, if no value is found. I am not sure whether this makes sense in your case.

---

I think we don't need a boolean here as we have a clear logic that we follow
1. ci value specified --> use that 
2. API key given --> use API 
3. If that fails or key not provided --> use CI for location from fallback table 
4. If not found or location not set --> use global fallback value from table

---

Yes, I implemented it as a potential way for the user to catch missing locations and more. I.e. the user could set the flag `fallback=false` and then the TDP would break if the model is not found / the CI would break if the location is not found. In every case if we go back to the worst value the program will then throw an error instead of a warning.

But this could be addressed in another PR or even in a future release.